ros2_demo.py

```diff
@@ -0,0 +1,362 @@
+        time_sec = self._time_input_port.Eval(context)
+        body_poses = self._poses_input_port.Eval(context)
+
+        ros_time = rclpy.time.Time(seconds=time_sec)
+
+        transforms = []
+        for joint in joints:
+            # Get poses of bodies on both sides of the joint
+            parent = joint.parent_body()
+            child = joint.child_body()
+            parent_pose = body_poses[int(parent.index())]
+            child_pose = body_poses[int(child.index())]
+
+            # Calculate the pose of child relative to parent
+            drake_tf = parent_pose.inverse().multiply(child_pose)
+
+            # Convert that into a TF transform ROS message
+            tf_stamped = TransformStamped()
+            tf_stamped.header.stamp = ros_time.to_msg()
+            tf_stamped.header.frame_id = parent.body_frame().name()
+            tf_stamped.child_frame_id = child.body_frame().name()
+            translation = tf_stamped.transform.translation
+            translation.x, translation.y, translation.z = \
+                drake_tf.translation()
+            rotation = tf_stamped.transform.rotation
+            rotation.w, rotation.x, rotation.y, rotation.z = \
+                drake_tf.rotation().ToQuaternion().wxyz()
+
+            transforms.append(tf_stamped)
+
+        # Publish transforms over a ROS topic!
+        tf_broadcaster.sendTransform(transforms)
+
+
+class JointStatePublisher(LeafSystem):
+
+    def __init__(self, *, publisher, joints, period_sec=1./60):
+        super().__init__()
+
+        # System will publish transforms for these joints only
+        self._joints = tuple(joints)
+
+        if 0 == len(self._joints):
+            raise ValueError('Need at least one joint to publish joint states')
+
+        for joint in self._joints:
+            if not Joint_.is_subclass_of_instantiation(type(joint)):
+                raise TypeError('joints must be an iterable of Joint_[T]')
+
+        # This event publishes TF transforms at a regular interval
+        self.DeclarePeriodicEvent(
+            period_sec=period_sec,
+            offset_sec=0.,
+            event=PublishEvent(
+                trigger_type=TriggerType.kPeriodic,
+                callback=self._publish_joint_states))
+
+    def _publish_joint_states(self, context, event):
+        # TODO(sloretz) How to get plant context for calculating joint angle?
+        # joint.get_angle(plant_context)
+        pass
+
+
+if __name__ == '__main__':
+
+    this_dir = os.path.abspath(os.path.dirname(__file__))
+
+    # sdf_file_path = FindResourceOrThrow(
+    #     "drake/manipulation/models/iiwa_description/iiwa7/"
+    #     "iiwa7_no_collision.sdf")
+    # print(sdf_file_path)
+
+    sdf_file_path = os.path.join(this_dir, 'ur10.sdf')
+
+    with open(os.path.join(this_dir, 'ur10.sdf.in'), 'r') as file_in:
+        with open(sdf_file_path, 'w') as file_out:
+            file_out.write(file_in.read().replace('PWD_GOES_HERE', this_dir))
```

Comment:
The demo uses `file` URI's to allow RViz and drake to find the meshes. This is rewriting the path so the demo is relocatable.